### PR TITLE
Added DateOnly overload methods to IsoWeek

### LINKED ISSUE #101597: [API Proposal]: DateOnly support in IsoWeek
### Background and motivation

At the moment the IsoWeek class only supports DateTime structs. 

Since the IsoWeek class is primarily (if not exclusively) based around dates rather than times, it would make sense to have it support the recently added DateOnly struct.

### API Proposal

Adding DateOnly variants of the public methods on the IsoWeek class.

```csharp
namespace System.Globalization
{
    public static class ISOWeek
    {
        //Added methods
        public static int GetWeekOfYear(DateOnly date);
        public static int GetYear(DateOnly date);
        public static DateOnly ToDateOnly(int year, int week, DayOfWeek dayOfWeek);
    }
}
```

### API Usage

```csharp
var week = IsoWeek.GetWeekOfYear(new DateOnly(2024, 4, 26));
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by leonluc-dev**
Fixes #101597

As discussed in proposal #101597

### CODE REVIEW

**Comment by leonluc-dev** [comment]
@dotnet-policy-service agree company="TimeTell B.V."

**Comment by huoyaoyuan** [comment]
To expose a public API, you also need to update the ref source file here: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime/ref/System.Runtime.cs

See also the documentation at https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md .

File: src/libraries/System.Private.CoreLib/src/System/Globalization/ISOWeek.cs [hunk 1]
```diff
@@ -34,6 +34,8 @@ public static int GetWeekOfYear(DateTime date)
             return week;
         }
 
+        public static int GetWeekOfYear(DateOnly date) => GetWeekOfYear(date.GetEquivalentDateTime());
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/ISOWeek.cs:37]: Any new public new API need to have the Xml triple-slashes docs on it. Please add them to all new APIs.

**Comment by leonluc-dev** [comment]
@huoyaoyuan @tarekgh 

Requested changes have been made.

**Comment by tarekgh** [comment]
Could you please add some tests in https://github.com/dotnet/runtime/blob/81efcade9e543a2c03457b3bc87a2bc76b36057e/src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs#L10 too? Thanks!

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs [hunk 2]
```diff
@@ -187,6 +244,25 @@ public DateData(DateTime date, int year, int week, DayOfWeek dayOfWeek)
             }
         }
 
+        private struct DateData
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:247]: Isn't already defined above?

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs [hunk 3]
```diff
@@ -132,6 +165,30 @@ public static void ToDateTime_WithInvalidDayOfWeek_Throws(int dayOfWeek)
...
+        }
+
+        [Theory]
+        [InlineData(-1)]
+        [InlineData(8)]
+        public static void ToDateOnly_WithInvalidDayOfWeek_Throws(int dayOfWeek)
+        {
+            AssertExtensions.Throws<ArgumentOutOfRangeException>(nameof(dayOfWeek), () => ISOWeek.ToDateOnly(2018, 1, (DayOfWeek)dayOfWeek));
+        }
+
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:191]: Can these be combined with the already existing methods?

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs (hunk 2 above)
Comment (leonluc-dev) [reply src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs]: Apologies, this was a diff error in the last commit. It should be fixed now.

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs (hunk 3 above)
Comment (leonluc-dev) [reply src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:191]: Calling both the ToDateTime and ToDateOnly checks in a single test method should be possible. I will adjust the tests to combine these.

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs (hunk 2 above)
Comment (tarekgh) [reply src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs]: No worries at all. Thanks for fixing it!

**Comment by leonluc-dev** [comment]
@tarekgh Consolidated the `ToDateTime`/`ToDateOnly` tests as requested

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs [hunk 4]
```diff
@@ -91,29 +106,48 @@
~        }
~
~        [Theory, MemberData(nameof(GetWeekOfYear_TestData))]
~        public static void GetWeekOfYear(DateTime date, int expected)
~        {
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:110]: >GetWeekOfYear [](http://example.com/codeflow?start=27&length=13)

xUnit doesn't allow using two test methods with the same name. Please rename this one or even merge it the two methods. This apply to GetYear too.

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs [hunk 5]
```diff
@@ -96,24 +111,43 @@ public static void GetWeekOfYear(DateTime date, int expected)
             Assert.Equal(expected, ISOWeek.GetWeekOfYear(date));
         }
 
+        [Theory, MemberData(nameof(GetWeekOfYear_DateOnly_TestData))]
+        public static void GetWeekOfYear(DateOnly date, int expected)
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:115]:
```suggest
        public static void GetWeekOfYearUsingDateOnly (DateOnly date, int expected)
````

Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:115]:
```suggest
        public static void GetWeekOfYearUsingDateOnly(DateOnly date, int expected)
```

Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:115]:
```suggestion
        public static void GetWeekOfYearUsingDateOnly(DateOnly date, int expected)
```

File: src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs [hunk 6]
```diff
@@ -96,24 +111,43 @@ public static void GetWeekOfYear(DateTime date, int expected)
...
+        }
+
         [Theory, MemberData(nameof(GetYear_TestData))]
         public static void GetYear(DateTime date, int expected)
         {
             Assert.Equal(expected, ISOWeek.GetYear(date));
         }
 
+        [Theory, MemberData(nameof(GetYear_DateOnly_TestData))]
+        public static void GetYear(DateOnly date, int expected)
```
Comment (tarekgh) [inline src/libraries/System.Runtime/tests/System.Globalization.Calendars.Tests/ISOWeek/ISOWeekTests.cs:127]:
```suggestion
        public static void GetYearUsingDateOnly(DateOnly date, int expected)
```

**Review by tarekgh — APPROVED** [review]
Thanks @leonluc-dev for helping with this change!

**Pull request merged**